File: source/src/test/java/org/rss/tools/mpl/grammar/LexerTest.java
package org.rss.tools.mpl.grammar;

import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.atn.ATNConfigSet;
import org.antlr.v4.runtime.dfa.DFA;
import org.junit.jupiter.api.Test;
import org.rss.tools.mpl.parsing.grammar2.MplLexer;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.BitSet;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;
import static org.rss.tools.mpl.parsing.grammar2.MplLexer.VOCABULARY;

public class LexerTest {

    final ANTLRErrorListener errorListener = new ANTLRErrorListener() {
        @Override
        public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
            System.err.format("Syntax Error (at line %d): %s : %s %n", line, msg, offendingSymbol);
        }

        @Override
        public void reportAmbiguity(Parser recognizer, DFA dfa, int startIndex, int stopIndex, boolean exact, BitSet ambigAlts, ATNConfigSet configs) {
            System.err.format("Ambiguity: %s - %s %n", dfa.toLexerString(), recognizer.getCurrentToken().getText());
        }

        @Override
        public void reportAttemptingFullContext(Parser recognizer, DFA dfa, int startIndex, int stopIndex, BitSet conflictingAlts, ATNConfigSet configs) {
        }

        @Override
        public void reportContextSensitivity(Parser recognizer, DFA dfa, int startIndex, int stopIndex, int prediction, ATNConfigSet configs) {
        }
    };
    
    private List<Token> getTokensFromText(String txt) throws IOException {
        InputStream is = new ByteArrayInputStream(txt.getBytes());
        CharStream cStream = CharStreams.fromStream(is);
        MplLexer lex = new MplLexer(cStream);
        lex.addErrorListener(errorListener);
        CommonTokenStream tokenStream = new CommonTokenStream(lex);
//        tokenStream.consume();
        tokenStream.fill();
        return tokenStream.getTokens();
    }

    @Test
    void testLexer() throws IOException {
        List<Token> tokens = getTokensFromText("document { \n function(param=\"valor\") \n }");

        System.out.println("Tokens: ");
        printTokens(tokens);
    }

    @Test
    void testFunction() throws IOException {
        List<Token> tokens = getTokensFromText("function ()");
        printTokens(tokens);

        assertEquals(4, tokens.size()); // inclues EOF
        assertEquals(MplLexer.WORD, tokens.get(0).getType());
        assertEquals(MplLexer.OPEN_PAR, tokens.get(1).getType());
        assertEquals(MplLexer.CLOSE_PAR, tokens.get(2).getType());


    }

    @Test
    void testParameters() throws IOException {
        List<Token> tokens = getTokensFromText("param = \"value1\", myNum = 5");
//        printTokens(tokens);
        assertAll(
                () -> assertEquals(8, tokens.size()),
                () -> assertEquals(MplLexer.WORD, tokens.get(0).getType()),
                () -> assertEquals(MplLexer.STRING, tokens.get(2).getType()),
                () -> assertEquals(MplLexer.NUMBER, tokens.get(6).getType(), VOCABULARY.getDisplayName(tokens.get(5).getType()))
        );
    }

    @Test
    void testText() throws IOException {
        List<Token> tokens = getTokensFromText("\"my text with spaces. & + symbols\", \"my other text with \n newLine\"");
        printTokens(tokens);

        assertAll(
                () -> assertEquals(4, tokens.size()),
                () -> assertEquals(MplLexer.STRING, tokens.get(0).getType())
//                () -> assertEquals(MplLexer.STRING, tokens.get(2).getType()),
//                () -> assertEquals(MplLexer.NUMBER, tokens.get(6).getType(), VOCABULARY.getDisplayName(tokens.get(5).getType()))
        );
    }

    private void printTokens(List<Token> tokens) {
        tokens.stream().map(t -> {
            StringBuilder sb = new StringBuilder();
            sb/*.append(t.getType())
                    .append(": ")*/
                    .append(VOCABULARY.getDisplayName(t.getType()))
                    .append("[")
                    .append(t.getText())
                    .append("] ");
            return sb.toString();
        }).forEach(System.out::print);
    }
}

File: source/src/test/kotlin/org/rss/tools/mpl/test/LexerTest.kt
package org.rss.tools.mpl.test

import org.antlr.v4.runtime.*
import org.antlr.v4.runtime.atn.ATNConfigSet
import org.antlr.v4.runtime.dfa.DFA
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import org.rss.tools.mpl.parsing.grammar2.MplLexer
import org.rss.tools.mpl.parsing.grammar2.MplLexer.VOCABULARY
import java.io.ByteArrayInputStream
import java.io.IOException
import java.util.*

class LexerTest {

    val errorListener: ANTLRErrorListener = object : ANTLRErrorListener {
        override fun syntaxError(recognizer: Recognizer<*, *>, offendingSymbol: Any, line: Int, charPositionInLine: Int, msg: String, e: RecognitionException) {
            System.err.format("Syntax Error (at line %d): %s : %s %n", line, msg, offendingSymbol)
        }

        override fun reportAmbiguity(recognizer: Parser, dfa: DFA, startIndex: Int, stopIndex: Int, exact: Boolean, ambigAlts: BitSet, configs: ATNConfigSet) {
            System.err.format("Ambiguity: %s - %s %n", dfa.toLexerString(), recognizer.currentToken.text)
        }

        override fun reportAttemptingFullContext(recognizer: Parser, dfa: DFA, startIndex: Int, stopIndex: Int, conflictingAlts: BitSet, configs: ATNConfigSet) {}

        override fun reportContextSensitivity(recognizer: Parser, dfa: DFA, startIndex: Int, stopIndex: Int, prediction: Int, configs: ATNConfigSet) {}
    }

    /** Boilerplate to process a string and generate the Tokens for it */
    @Throws(IOException::class)
    private fun getTokensFromText(txt: String): List<Token> {
        val iStream = ByteArrayInputStream(txt.toByteArray())
        val cStream = CharStreams.fromStream(iStream)
        val lex = MplLexer(cStream)
        lex.addErrorListener(errorListener)
        val tokenStream = CommonTokenStream(lex)
        //        tokenStream.consume();
        tokenStream.fill()
        return tokenStream.tokens
    }

    @Test
    @Throws(IOException::class)
    internal fun testLexer() {
        val tokens = getTokensFromText("document { \n function(param=\"valor\") \n }")

        assertTrue(tokens.stream().anyMatch { p -> p.type == MplLexer.OPEN_BRACE })
        println("Tokens: ")
        printTokens(tokens)
    }

    @Test
    @Throws(IOException::class)
    internal fun testFunction() {
        val tokens = getTokensFromText("function ()")
        printTokens(tokens)

        assertEquals(4, tokens.size) // includes EOF
        assertEquals(MplLexer.WORD, tokens[0].type)
        assertEquals(MplLexer.OPEN_PAR, tokens[1].type)
        assertEquals(MplLexer.CLOSE_PAR, tokens[2].type)
    }

    @Test
    @Throws(IOException::class)
    internal fun testParameters() {
        val tokens = getTokensFromText("param = \"value1\", myNum = 5")
        //        printTokens(tokens);
        assertAll(
                Executable { assertEquals(8, tokens.size) },
                Executable { assertEquals(MplLexer.WORD, tokens[0].type) },
                Executable { assertEquals(MplLexer.STRING, tokens[2].type) },
                Executable { assertEquals(MplLexer.NUMBER, tokens[6].type, VOCABULARY.getDisplayName(tokens[5].type)) }
        )
    }

    @Test
    @Throws(IOException::class)
    internal fun testText() {
        val tokens = getTokensFromText("\"my text with spaces. & + symbols\", \"my other text with \n newLine\"")
        printTokens(tokens)

        assertAll(
                Executable { assertEquals(4, tokens.size) },
                Executable { assertEquals(MplLexer.STRING, tokens[0].type) }
        )
    }

    private fun printTokens(tokens: List<Token>) {
        tokens.stream().map { t ->
            "${VOCABULARY.getDisplayName(t.type)} [${t.text}] "
        }.forEach { print(it) }
    }
}
